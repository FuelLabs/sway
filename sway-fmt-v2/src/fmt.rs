use crate::utils::{
    indent_style::Shape, newline_style::apply_newline_style, program_type::insert_program_type,
};
use std::{path::Path, sync::Arc};
use sway_core::BuildConfig;

pub use crate::{
    config::manifest::Config,
    error::{ConfigError, FormatterError},
};

#[derive(Debug, Default)]
pub struct Formatter {
    pub shape: Shape,
    pub config: Config,
}

pub type FormattedCode = String;

pub trait Format {
    fn format(
        &self,
        formatted_code: &mut FormattedCode,
        formatter: &mut Formatter,
    ) -> Result<(), FormatterError>;
}

impl Formatter {
    pub fn from_dir(dir: &Path) -> Result<Self, ConfigError> {
        let config = match Config::from_dir(dir) {
            Ok(config) => config,
            Err(ConfigError::NotFound) => Config::default(),
            Err(e) => return Err(e),
        };
        let shape = Shape::default();
        Ok(Self { config, shape })
    }
    pub fn format(
        &mut self,
        src: Arc<str>,
        build_config: Option<&BuildConfig>,
    ) -> Result<FormattedCode, FormatterError> {
        let path = build_config.map(|build_config| build_config.canonical_root_module());
        let src_len = src.len();
        let module = sway_parse::parse_file(src, path)?;
        // Get parsed items
        let items = module.items;
        // Get the program type (script, predicate, contract or library)
        let program_type = module.kind;

        // Formatted code will be pushed here with raw newline stlye.
        // Which means newlines are not converted into system-specific versions until `apply_newline_style()`.
        // Use the length of src as a hint of the memory size needed for `raw_formatted_code`,
        // which will reduce the number of reallocations
        let mut raw_formatted_code = String::with_capacity(src_len);

        // Insert program type to the formatted code.
        insert_program_type(&mut raw_formatted_code, program_type)?;

        // Insert parsed & formatted items into the formatted code.
        let mut iter = items.iter().peekable();
        while let Some(item) = iter.next() {
            // format Annotated<ItemKind>
            item.format(&mut raw_formatted_code, self)?;
            if iter.peek().is_some() {
                raw_formatted_code.push('\n');
            }
        }

        let mut formatted_code = String::from(&raw_formatted_code);
        // Replace newlines with specified `NewlineStyle`
        apply_newline_style(
            self.config.whitespace.newline_style,
            &mut formatted_code,
            &raw_formatted_code,
        )?;
        if formatted_code.ends_with('\n') {
            formatted_code.pop();
        }

        Ok(formatted_code)
    }
}

#[cfg(test)]
mod tests {
    use super::Formatter;
    use crate::config::user_def::FieldAlignment;
    use std::sync::Arc;

    #[test]
    fn test_const() {
        let sway_code_to_format = r#"contract;
pub const TEST:u16=10;"#;
        let correct_sway_code = r#"contract;

pub const TEST: u16 = 10;"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_struct_multiline_line_alignment() {
        let sway_code_to_format = r#"contract;
pub struct Foo<T, P> {
   barbazfoo: u64,
   baz  : bool,
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo<T, P> {
    barbazfoo : u64,
    baz       : bool,
}"#;

        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(40);
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_struct_single_line() {
        let sway_code_to_format = r#"contract;
pub struct Foo {
    bar: u64,
    baz: bool,
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo { bar: u64, baz: bool }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_enum_single_line() {
        let sway_code_to_format = r#"contract;
pub enum Foo {
    bar: u64,
    baz: bool,
}
"#;
        let correct_sway_code = r#"contract;

pub enum Foo { bar: u64, baz: bool }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_struct_multi_line() {
        let sway_code_to_format = r#"contract;
pub struct Foo {
    bar: u64,
    baz: bool
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo {
    bar: u64,
    baz: bool,
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_enum_without_variant_alignment() {
        let sway_code_to_format = r#"contract;

enum Color {
    Blue: (), Green: (),
            Red: (),
    Silver: (),
                    Grey: () }
        "#;
        let correct_sway_code = r#"contract;

enum Color {
    Blue: (),
    Green: (),
    Red: (),
    Silver: (),
    Grey: (),
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_enum_with_variant_alignment() {
        let sway_code_to_format = r#"contract;

enum Color {
    Blue: (), Green: (),
            Red: (),
    Silver: (),
                    Grey: (), }
        "#;
        let correct_sway_code = r#"contract;

enum Color {
    Blue   : (),
    Green  : (),
    Red    : (),
    Silver : (),
    Grey   : (),
}"#;

        // Creating a config with enum_variant_align_threshold that exceeds longest variant length
        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(20);

        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_item_abi_with_generics_and_attributes() {
        let sway_code_to_format = r#"contract;

abi StorageMapExample {
    #[storage(write)]fn insert_into_map1(key: u64, value: u64);

fn hello(key: u64, value: u64);
}"#;
        let correct_sway_code = r#"contract;

abi StorageMapExample {
    #[storage(write)]
    fn insert_into_map1(key: u64, value: u64);

    fn hello(key: u64, value: u64);
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_multi_items() {
        let sway_code_to_format = r#"contract;

pub const TEST: u16 = 10;
pub const TEST1: u16 = 10;"#;
        let correct_sway_code = r#"contract;

pub const TEST: u16 = 10;
pub const TEST1: u16 = 10;"#;

        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_ty_formatting() {
        let sway_code_to_format = r#"contract;

enum TestTy {
    Infer:
    _,
    Array : [u8;
    40],
    String:         str[
    4
    ],
    PathType     : root::
example::
    type,
    TupleNil: (),
    Tuple: (   u64, 
        u32
    ),
}"#;
        let correct_sway_code = r#"contract;

enum TestTy {
    Infer: _,
    Array: [u8; 40],
    String: str[4],
    PathType: root::example::type,
    TupleNil: (),
    Tuple: (u64, u32),
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code);
    }
    fn test_storage_without_alignment() {
        let sway_code_to_format = r#"contract;

storage{foo:Test=Test{},bar
: 
    Test=Test{}
, baz: u64 } 
"#;
        let correct_sway_code = r#"contract;

storage {
    foo: Test,
    bar: Test,
    baz: u64,
}"#;

        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_storage_with_alignment() {
        let sway_code_to_format = r#"contract;

storage {
 long_var_name: Type1=Type1{},
      var2: Type2=Type2{},
}
"#;
        let correct_sway_code = r#"contract;

storage {
    long_var_name : Type1,
    var2          : Type2,
}"#;

        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(50);
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_storage_single_line() {
        let sway_code_to_format = r#"contract;

storage {
 long_var_name: Type1=Type1{},
      var2: Type2=Type2{},
}
"#;
        let correct_sway_code = r#"contract;

storage { long_var_name: Type1, var2: Type2 }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_item_fn() {
        let sway_code_to_format = r#"contract;

pub fn hello( person: String ) -> String {let greeting = 42;greeting.to_string()}
fn goodbye() -> usize {let farewell: usize = 5; farewell }"#;
        let correct_sway_code = r#"contract;

pub fn hello(person: String) -> String {
    let greeting = 42;
    greeting.to_string()
}

fn goodbye() -> usize {
    let farewell: usize = 5;
    farewell
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_same_line_where() {
        let sway_code_to_format = r#"contract;

pub fn hello( person: String ) -> String where T: Eq,{let greeting = 42;greeting.to_string()}"#;
        let correct_sway_code = r#"contract;

pub fn hello(person: String) -> String
where
    T: Eq,
{
    let greeting = 42;
    greeting.to_string()
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_trait_and_super_trait() {
        let sway_code_to_format = r#"library traits;

trait Person{ fn name( self )->String;fn age( self )->usize; }
trait Student:Person {fn university(self) -> String;}
trait Programmer {fn fav_language(self) -> String;}
trait CompSciStudent: Programmer+Student {fn git_username(self) -> String;}"#;
        let correct_sway_code = r#"library traits;

trait Person {
    fn name(self) -> String;

    fn age(self) -> usize;
}

trait Student: Person {
    fn university(self) -> String;
}

trait Programmer {
    fn fav_language(self) -> String;
}

trait CompSciStudent: Programmer + Student {
    fn git_username(self) -> String;
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
}
