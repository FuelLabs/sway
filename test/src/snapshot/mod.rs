use anyhow::Result;
use libtest_mimic::{Arguments, Trial};
use normalize_path::NormalizePath;
use regex::Regex;
use std::{
    path::{Path, PathBuf},
    str::FromStr,
    sync::Once,
};

static FORC_COMPILATION: Once = Once::new();
static FORC_DOC_COMPILATION: Once = Once::new();

fn compile_forc() {
    let args = vec!["b", "--release", "-p", "forc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

fn compile_forc_doc() {
    let args = vec!["b", "--release", "-p", "forc-doc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

pub(super) async fn run(filter_regex: Option<&regex::Regex>) -> Result<()> {
    let repo_root = PathBuf::from_str(&std::env::var("CARGO_MANIFEST_DIR")?)?
        .parent()
        .unwrap()
        .to_path_buf();
    let test_root = repo_root.join("test");
    let test_programs_dir = test_root.join("src/e2e_vm_tests/test_programs/");

    let args = Arguments {
        filter: filter_regex.as_ref().map(|filter| filter.to_string()),
        nocapture: true,
        ..Default::default()
    };

    let tests = discover_tests(&test_root)
        .into_iter()
        .map(|dir| {
            let name = dir
                .strip_prefix(&test_programs_dir)
                .unwrap()
                .display()
                .to_string();

            let repo_root = repo_root.clone();
            Trial::test(name, move || {
                let snapshot_toml =
                    std::fs::read_to_string(format!("{}/snapshot.toml", dir.display()))?;
                let snapshot_toml = toml::from_str::<toml::Value>(&snapshot_toml)?;
                let cmds = if let Some(cmds) = snapshot_toml.get("cmds") {
                    cmds.as_array()
                        .unwrap()
                        .iter()
                        .map(|v| v.as_str().unwrap())
                        .collect::<Vec<_>>()
                } else {
                    vec!["forc build --path {root}"]
                };

                let root = dir.strip_prefix(&repo_root).unwrap().display().to_string();

                use std::fmt::Write;
                let mut snapshot = String::new();

                for cmd in cmds {
                    let cmd = cmd.replace("{root}", &root);

                    let _ = writeln!(&mut snapshot, "> {}", cmd);

                    let mut last_output: Option<String> = None;

                    for cmd in cmd.split("|") {
                        let cmd = cmd.trim();

                        // known commands
                        let cmd = if let Some(cmd) = cmd.strip_prefix("forc doc ") {
                            FORC_DOC_COMPILATION.call_once(|| {
                                compile_forc_doc();
                            });
                            format!("target/release/forc-doc {cmd} 1>&2")
                        } else if let Some(cmd) = cmd.strip_prefix("forc ") {
                            FORC_COMPILATION.call_once(|| {
                                compile_forc();
                            });
                            format!("target/release/forc {cmd} 1>&2")
                        } else if let Some(cmd) = cmd.strip_prefix("grep ") {
                            let arg = cmd.trim();
                            if let Some(l) = last_output.take() {
                                let mut new_output = String::new();
                                for line in l.lines() {
                                    if line.contains(arg) {
                                        new_output.push_str(line);
                                        new_output.push('\n');
                                    }
                                }
                                last_output = Some(new_output);
                            }
                            continue;
                        } else {
                            panic!("Not supported. Possible commands: forc")
                        };

                        let o = duct::cmd!("bash", "-c", cmd.clone())
                            .dir(repo_root.clone())
                            .stderr_to_stdout()
                            .stdout_capture();

                        let o = if let Some(last_output) = last_output.as_ref() {
                            o.stdin_bytes(last_output.as_bytes())
                        } else {
                            o
                        };

                        let o = o.env("COLUMNS", "10").unchecked().start().unwrap();
                        let o = o.wait().unwrap();
                        last_output = Some(clean_output(&format!(
                            "exit status: {}\noutput:\n{}",
                            o.status.code().unwrap(),
                            std::str::from_utf8(&o.stdout).unwrap(),
                        )));
                    }

                    let _ = writeln!(&mut snapshot, "{}", last_output.unwrap_or_default());
                }

                fn stdout(root: &str, snapshot: &str) {
                    let root = PathBuf::from_str(root).unwrap();
                    let root = root.normalize();

                    let mut insta = insta::Settings::new();
                    insta.set_snapshot_path(root);
                    insta.set_prepend_module_to_snapshot(false);
                    insta.set_omit_expression(true);
                    let scope = insta.bind_to_scope();
                    insta::assert_snapshot!("stdout", snapshot);
                    drop(scope);
                }
                stdout(&format!("{}/{root}", repo_root.display()), &snapshot);

                Ok(())
            })
        })
        .collect();

    libtest_mimic::run(&args, tests).exit();
}

pub fn discover_tests(test_root: &Path) -> Vec<PathBuf> {
    use glob::glob;

    let mut entries = vec![];

    let pattern = format!("{}/**/snapshot.toml", test_root.display());
    for entry in glob(&pattern)
        .expect("Failed to read glob pattern")
        .flatten()
    {
        entries.push(entry.parent().unwrap().to_owned())
    }

    entries
}

fn clean_output(output: &str) -> String {
    #[derive(Default)]
    struct RawText(String);

    impl vte::Perform for RawText {
        fn print(&mut self, c: char) {
            self.0.push(c);
        }

        fn execute(&mut self, _: u8) {}

        fn hook(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn put(&mut self, b: u8) {
            self.0.push(b as char);
        }

        fn unhook(&mut self) {}

        fn osc_dispatch(&mut self, _: &[&[u8]], _: bool) {}

        fn csi_dispatch(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn esc_dispatch(&mut self, _: &[u8], _: bool, _: u8) {}
    }

    let mut raw = String::new();
    for line in output.lines() {
        let mut performer = RawText::default();
        let mut p = vte::Parser::new();
        for b in line.as_bytes() {
            p.advance(&mut performer, *b);
        }
        raw.push_str(&performer.0);
        raw.push('\n');
    }

    // Remove absolute paths from snapshot tests
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let manifest_dir: PathBuf = PathBuf::from(manifest_dir);
    let parent = manifest_dir.parent().unwrap();
    let result = raw.replace(&format!("{}/", parent.display()), "");

    // Remove compilation time
    let r = Regex::new("(Finished (debug|release) \\[.*?\\] target\\(s\\) \\[.*?\\] in )(.*?s)")
        .unwrap();
    let result = r.replace(&result, "$1???");

    result.to_string()
}
