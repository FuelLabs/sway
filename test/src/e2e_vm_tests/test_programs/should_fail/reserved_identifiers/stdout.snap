---
source: test/src/snapshot/mod.rs
---
> forc build --path test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers --release --experimental error_type
exit status: 1
output:
    Building test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers
   Compiling library std (test/src/e2e_vm_tests/reduced_std_libs/sway-lib-std-core)
   Compiling script reserved_identifiers (test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers)
warning
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:33:13
   |
31 |     let mut where = 0;
32 |     let mut ref = 0;
33 |     let mut deref = 0;
   |             ----- This declaration is never used.
34 |     let mut true = 0;
35 |     let mut false = 0;
   |
____

error
 --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:4:13
  |
2 | 
3 | fn main() {
4 |     let mut script = 0;
  |             ^^^^^^ Identifier cannot be a reserved keyword.
5 |     let mut contract = 0;
6 |     let mut predicate = 0;
  |
____

error
 --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:5:13
  |
3 | fn main() {
4 |     let mut script = 0;
5 |     let mut contract = 0;
  |             ^^^^^^^^ Identifier cannot be a reserved keyword.
6 |     let mut predicate = 0;
7 |     let mut library = 0;
  |
____

error
 --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:6:13
  |
4 |     let mut script = 0;
5 |     let mut contract = 0;
6 |     let mut predicate = 0;
  |             ^^^^^^^^^ Identifier cannot be a reserved keyword.
7 |     let mut library = 0;
8 |     let mut mod = 0;
  |
____

error
 --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:7:13
  |
5 |     let mut contract = 0;
6 |     let mut predicate = 0;
7 |     let mut library = 0;
  |             ^^^^^^^ Identifier cannot be a reserved keyword.
8 |     let mut mod = 0;
9 |     let mut pub = 0;
  |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:8:13
   |
 6 |     let mut predicate = 0;
 7 |     let mut library = 0;
 8 |     let mut mod = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
 9 |     let mut pub = 0;
10 |     let mut use = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:9:13
   |
 7 |     let mut library = 0;
 8 |     let mut mod = 0;
 9 |     let mut pub = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
10 |     let mut use = 0;
11 |     let mut as = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:10:13
   |
 8 |     let mut mod = 0;
 9 |     let mut pub = 0;
10 |     let mut use = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
11 |     let mut as = 0;
12 |     let mut struct = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:11:13
   |
 9 |     let mut pub = 0;
10 |     let mut use = 0;
11 |     let mut as = 0;
   |             ^^ Identifier cannot be a reserved keyword.
12 |     let mut struct = 0;
13 |     let mut enum = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:12:13
   |
10 |     let mut use = 0;
11 |     let mut as = 0;
12 |     let mut struct = 0;
   |             ^^^^^^ Identifier cannot be a reserved keyword.
13 |     let mut enum = 0;
14 |     let mut self = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:13:13
   |
11 |     let mut as = 0;
12 |     let mut struct = 0;
13 |     let mut enum = 0;
   |             ^^^^ Identifier cannot be a reserved keyword.
14 |     let mut self = 0;
15 |     let mut fn = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:14:13
   |
12 |     let mut struct = 0;
13 |     let mut enum = 0;
14 |     let mut self = 0;
   |             ^^^^ Identifier cannot be a reserved keyword.
15 |     let mut fn = 0;
16 |     let mut trait = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:15:13
   |
13 |     let mut enum = 0;
14 |     let mut self = 0;
15 |     let mut fn = 0;
   |             ^^ Identifier cannot be a reserved keyword.
16 |     let mut trait = 0;
17 |     let mut impl = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:16:13
   |
14 |     let mut self = 0;
15 |     let mut fn = 0;
16 |     let mut trait = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
17 |     let mut impl = 0;
18 |     let mut for = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:17:13
   |
15 |     let mut fn = 0;
16 |     let mut trait = 0;
17 |     let mut impl = 0;
   |             ^^^^ Identifier cannot be a reserved keyword.
18 |     let mut for = 0;
19 |     let mut abi = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:18:13
   |
16 |     let mut trait = 0;
17 |     let mut impl = 0;
18 |     let mut for = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
19 |     let mut abi = 0;
20 |     let mut const = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:19:13
   |
17 |     let mut impl = 0;
18 |     let mut for = 0;
19 |     let mut abi = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
20 |     let mut const = 0;
21 |     let mut storage = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:20:13
   |
18 |     let mut for = 0;
19 |     let mut abi = 0;
20 |     let mut const = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
21 |     let mut storage = 0;
22 |     let mut str = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:21:13
   |
19 |     let mut abi = 0;
20 |     let mut const = 0;
21 |     let mut storage = 0;
   |             ^^^^^^^ Identifier cannot be a reserved keyword.
22 |     let mut str = 0;
23 |     let mut asm = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:22:13
   |
20 |     let mut const = 0;
21 |     let mut storage = 0;
22 |     let mut str = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
23 |     let mut asm = 0;
24 |     let mut return = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:23:13
   |
21 |     let mut storage = 0;
22 |     let mut str = 0;
23 |     let mut asm = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
24 |     let mut return = 0;
25 |     let mut if = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:24:13
   |
22 |     let mut str = 0;
23 |     let mut asm = 0;
24 |     let mut return = 0;
   |             ^^^^^^ Identifier cannot be a reserved keyword.
25 |     let mut if = 0;
26 |     let mut else = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:25:13
   |
23 |     let mut asm = 0;
24 |     let mut return = 0;
25 |     let mut if = 0;
   |             ^^ Identifier cannot be a reserved keyword.
26 |     let mut else = 0;
27 |     let mut match = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:26:13
   |
24 |     let mut return = 0;
25 |     let mut if = 0;
26 |     let mut else = 0;
   |             ^^^^ Identifier cannot be a reserved keyword.
27 |     let mut match = 0;
28 |     let mut mut = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:27:13
   |
25 |     let mut if = 0;
26 |     let mut else = 0;
27 |     let mut match = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
28 |     let mut mut = 0;
29 |     let mut let = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:28:13
   |
26 |     let mut else = 0;
27 |     let mut match = 0;
28 |     let mut mut = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
29 |     let mut let = 0;
30 |     let mut while = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:29:13
   |
27 |     let mut match = 0;
28 |     let mut mut = 0;
29 |     let mut let = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
30 |     let mut while = 0;
31 |     let mut where = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:30:13
   |
28 |     let mut mut = 0;
29 |     let mut let = 0;
30 |     let mut while = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
31 |     let mut where = 0;
32 |     let mut ref = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:31:13
   |
29 |     let mut let = 0;
30 |     let mut while = 0;
31 |     let mut where = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
32 |     let mut ref = 0;
33 |     let mut deref = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:32:13
   |
30 |     let mut while = 0;
31 |     let mut where = 0;
32 |     let mut ref = 0;
   |             ^^^ Identifier cannot be a reserved keyword.
33 |     let mut deref = 0;
34 |     let mut true = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:34:13
   |
32 |     let mut ref = 0;
33 |     let mut deref = 0;
34 |     let mut true = 0;
   |             ^^^^ Identifier cannot be a reserved keyword.
35 |     let mut false = 0;
36 |     let mut panic = 0;
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:35:13
   |
33 |     let mut deref = 0;
34 |     let mut true = 0;
35 |     let mut false = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
36 |     let mut panic = 0;
37 | }
   |
____

error
  --> test/src/e2e_vm_tests/test_programs/should_fail/reserved_identifiers/src/main.sw:36:13
   |
34 |     let mut true = 0;
35 |     let mut false = 0;
36 |     let mut panic = 0;
   |             ^^^^^ Identifier cannot be a reserved keyword.
37 | }
   |
____

  Aborting due to 32 errors.
error: Failed to compile reserved_identifiers
